python/ray/experimental/serve/kv_store_service.py

```diff
@@ -0,0 +1,174 @@
+import json
+from abc import ABC
+
+import ray
+import ray.experimental.internal_kv as ray_kv
+from ray.experimental.serve.utils import logger
+
+
+class NamespacedKVStore(ABC):
+    """Abstract base class for a namespaced key-value store.
+
+    The idea is that multiple key-value stores can be created while sharing
+    the same database connection. The keys of each instance are namespaced
+    for conflict avoidance.
+
+    Example:
+
+    >>> store_ns1 = NamespacedKVStore(namespace="ns1")
+    >>> store_ns2 = NamespacedKVStore(namespace="ns2")
+    # Two store can share the same connection like Redis or SQL Table
+    >>> store_ns1.put("same-key", 1)
+    >>> store_ns1.get("same-key")
+    1
+    >>> store_ns2.put("same-key", 2)
+    >>> store_ns2.get("same-key", 2)
+    2
+    """
+
+    def __init__(self, namespace):
+        raise NotImplementedError()
+
+    def get(self, key):
+        """Retrieve the value given key
+
+        Args:
+            key (str)
+        """
+        raise NotImplementedError()
+
+    def put(self, key, value):
+        """Store the correponding entry key -> value.
+
+        Args:
+            key (str)
+            value (object): any serializable object. The serialization method
+                is determined by the subclass implementation.
+        """
+        raise NotImplementedError()
+
+    def as_dict(self):
+        """Return the entire namespace as python dictionary.
+
+        Returns:
+            data (dict): key value pairs in current namespace
+        """
+        raise NotImplementedError()
+
+
+class InMemoryKVStore(NamespacedKVStore):
+    """A reference implementation"""
+
+    def __init__(self, namespace):
+        self.data = dict()
+
+        # namepsace is ignored because each namespace is backed by
+        # one in memory dictionary
+        self.namespace = namespace
+
+    def get(self, key):
+        return self.data[key]
+
+    def put(self, key, value):
+        self.data[key] = value
+
+    def as_dict(self):
+        return self.data.copy()
+
+
+class RayInternalKVStore(NamespacedKVStore):
+    """A NamespacedKVStore implementation using ray's `internal_kv`"""
+
+    def __init__(self, namespace):
+        assert ray_kv._internal_kv_initialized()
+        self.index_key = "RAY_SERVE_INDEX"
+        self.namespace = namespace
+        self._put(self.index_key, [])
+
+    def _format_key(self, key):
+        return "{ns}-{key}".format(ns=self.namespace, key=key)
+
+    def _remove_format_key(self, formatted_key):
+        return formatted_key.replace(self.namespace + "-", "")
+
+    def _serialize(self, obj):
+        return json.dumps(obj)
+
+    def _deserialize(self, buffer):
+        return json.loads(buffer)
+
+    def _put(self, key, value):
+        ray_kv._internal_kv_put(
+            self._format_key(self._serialize(key)),
+            self._serialize(value),
```

Comment:
Should the serialization be part of the KV client implementation or be separate?

Reply:
I'm ok with having serialization to be part of the KV client because different DB backend may require different serialization methods (custom ORM serialization for SQL). We can separate them out when they become redundant.